src/main/scala/com/rbmhtechnology/calliope/scaladsl/TransactionalEventProducer.scala

```diff
@@ -0,0 +1,257 @@
+      copy(kafkaProducerSettings = kafkaProducerSettings.withCloseTimeout(closeTimeout))
+
+    def withProducerDispatcher(dispatcher: String): Settings =
+      copy(kafkaProducerSettings = kafkaProducerSettings.withDispatcher(dispatcher))
+
+    def withProducerParallelism(parallelism: Int): Settings =
+      copy(kafkaProducerSettings = kafkaProducerSettings.withParallelism(parallelism))
+
+    def withProducerProperty(key: String, value: String): Settings =
+      copy(kafkaProducerSettings = kafkaProducerSettings.withProperty(key, value))
+
+    private def copy(readBufferSize: Int = readBufferSize,
+                     readInterval: FiniteDuration = readInterval,
+                     deleteInterval: FiniteDuration = deleteInterval,
+                     transactionTimeout: FiniteDuration = transactionTimeout,
+                     kafkaProducerSettings: ProducerSettings[String, SequencedMessage[AnyRef]] = kafkaProducerSettings): Settings =
+      new Settings(readBufferSize, readInterval, deleteInterval, transactionTimeout, kafkaProducerSettings)
+  }
+}
+
+class TransactionalEventProducer[A] private(sourceId: String, topic: String, eventStore: EventStore[A], onFailure: FailureHandler, settings: Settings)
+                                           (implicit val system: ActorSystem) extends IoDispatcher {
+
+  import ProducerGraphRunner._
+
+  private var running = false
+
+  private val graph = ProducerGraph(
+    EventStoreReader.withTimestampGapDetection(sourceId, eventStore, settings.transactionTimeout),
+    new EventStoreDeleter(eventStore),
+    settings)
+
+  private lazy val graphRunner = system.actorOf(Props(new GraphRunnerSupervisor(graph, onFailure)))
+
+  def run(): EventWriter[A] = {
+    running = true
+    val runner = graphRunner
+    new EventStoreWriter[A](topic, eventStore, _ => runner ! NotifyCommit)
+  }
+
+  def stop(timeout: FiniteDuration = 5.seconds): Future[Boolean] = {
+    if (running)
+      gracefulStop(graphRunner, timeout)
+    else
+      Future.successful(true)
+  }
+}
+
+private object ProducerGraph {
+  def apply[A](eventReader: EventReader[A], eventDeleter: EventDeleter, settings: Settings): ProducerGraph =
+    EventSource(eventReader, settings.readBufferSize, settings.readInterval)
+      .viaMat(ProducerFlow[EventRecord[A]].toMessage)(Keep.left)
+      .viaMat(Producer.flow(settings.producerSettings(), settings.producer()))(Keep.left)
+      .map(_.message.record.value())
+      .toMat(EventDeletion.sink(eventDeleter, settings.deleteInterval))(Keep.both)
+}
+
+private class GraphRunnerSupervisor(graph: ProducerGraph, onFailure: FailureHandler) extends Actor {
+  import SupervisorStrategyExtensions._
+
+  override def supervisorStrategy: SupervisorStrategy = OneForOneStrategy()(SupervisorStrategy.defaultDecider.foreach {
+    case (err, Stop) =>
+      onFailure.apply(err)
+    case (err, Escalate) =>
+      onFailure.apply(err)
+  })
```

Comment:
I originally called the `onFailure` in the `onStop` of the child, but had 2 problems with it
1) I was not able to get the cause of the stop
2) `onStop` is also triggered on a clean shutdown, so the `onFailure` handler would always be called, on stop of the `TransactionalEventProducer` which lead to a misleading behaviour

IMO `preRestart` cannot be used because `onFailure` should only be triggered if no restart is performed - the child is shutdown and doesn't work anymore.

Regarding exceptions in the handler - good input, these should be caught in the `foreach` method of the `SupervisorExtensions`.
WDYT?